lib/clone.js

```diff
@@ -0,0 +1,66 @@
+export const getFinalError = (initialError, message) => {
+	const error = createFinalError(initialError, message);
+	previousErrors.add(error);
+	return error;
+};
+
+const createFinalError = (error, message) => {
+	if (Object.prototype.toString.call(error) !== '[object Error]') {
+		return new Error(message);
+	}
+
+	return previousErrors.has(error)
+		? cloneError(error, message)
+		: setErrorMessage(error, message);
+};
+
+const cloneError = (oldError, newMessage) => {
+	const {name, message, stack} = oldError;
+	const error = new Error(newMessage);
+	error.stack = fixStack(name, stack, message, newMessage);
+	Object.defineProperty(error, 'name', {value: name, enumerable: false, configurable: true, writable: true});
+	copyErrorProperties(error, oldError);
+	return error;
+};
+
+const copyErrorProperties = (newError, previousError) => {
+	for (const propertyName of COPIED_ERROR_PROPERTIES) {
+		const descriptor = Object.getOwnPropertyDescriptor(previousError, propertyName);
+		if (descriptor !== undefined) {
+			Object.defineProperty(newError, propertyName, descriptor);
+		}
+	}
+};
+
+// Known error properties
+const COPIED_ERROR_PROPERTIES = [
+	'cause',
+	'errors',
+	'code',
+	'errno',
+	'syscall',
+	'path',
+	'dest',
+	'address',
+	'port',
+	'info',
+];
+
+// Sets `error.message`.
+// Fixes `error.stack` not being updated when it has been already accessed, since it is memoized by V8.
+const setErrorMessage = (error, newMessage) => {
+	const {name, message, stack} = error;
+	error.message = newMessage;
+	error.stack = fixStack(name, stack, message, newMessage);
+	return error;
+};
+
+const fixStack = (name, stack, message, newMessage) => stack.includes(newMessage)
+	? stack
+	: stack.replace(`${name}: ${message}`, `${name}: ${newMessage}`);
```

Comment:
This works with `--enable-source-maps`, which usually prepends a preview of the source code at the beginning of `error.stack`.